File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.py
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""

class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        
        # Create a new node for each original node and connect them
        current = head
        while current:
            new_node = Node(current.val, None)
            new_node.next = current.next
            current.next = new_node
            current = new_node.next
        
        # Set random pointers for the new nodes
        current = head
        while current:
            if current.random:
                current.next.random = current.random.next
            current = current.next.next
        
        # Separate the original and new linked lists
        current = head
        new_head = current.next
        while current:
            new_node = current.next
            current.next = new_node.next
            if new_node.next:
                new_node.next = new_node.next.next
            current = current.next
        
        return new_head
        

File: 0138-copy-list-with-random-pointer/0138-copy-list-with-random-pointer.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* next;
    Node* random;
    
    Node(int _val) {
        val = _val;
        next = NULL;
        random = NULL;
    }
};
*/

class Solution {
public:
    void putCopyinBetween(Node* head){
        Node* temp = head;
        while(temp != nullptr){
            Node* copyNode = new Node(temp->val);
            Node* nextNode = temp->next;
            copyNode->next = nextNode;
            temp->next = copyNode;
            temp = nextNode;
        }
    }
    void connectRandomPointer(Node* head){
        Node* temp = head;
        while(temp != nullptr){
            Node* copyNode = temp->next;
            if(temp->random != nullptr){
                copyNode->random = temp->random->next;
            }
            else{
                copyNode->random = nullptr;
            }
            temp = temp->next->next;
        }
    }
    Node* connectNextandExtract(Node* head){
        Node* temp = head;
        Node* dummy = new Node(-1);
        Node* res = dummy;
        while(temp != nullptr){
            res->next = temp->next;
            res = res->next;
            temp->next = temp->next->next;
            temp = temp->next;
        }
        return dummy->next;
    }
    Node* copyRandomList(Node* head) {
        putCopyinBetween(head);
        connectRandomPointer(head);
        return connectNextandExtract(head);
    }
};